tests/remote_resources/xtcp_ping_pong.py

```diff
@@ -0,0 +1,160 @@
+    while tests_to_perform:
+        time.sleep(args.delay_between_packets)
+        length_of_message = random.randint(1, 50)
+        # Don't allow an 'a' char to be sent, as this kills the remote device
+        message = ''.join( [random.choice(string.lowercase[1:]) for c in xrange(length_of_message)] )
+    
+        try:
+            sock.send(message)
+            returned_message = sock.recv(1460)
+            if returned_message != message:
+                failures.append('FAILURE: Mismatched packet\n' +
+                                '  Local port:    {}\n'.format(sock.getsockname()[1]) +
+                                '  Remote port:   {}\n'.format(process_port) +
+                                '  Packet length: {}\n'.format(length_of_message) +
+                                '  Recv length:   {}\n'.format(len(returned_message)) +
+                                '  Sent message:  \n{}\n'.format(format_message(message, 40)) +
+                                '  Recv message:  \n{}\n'.format(format_message(returned_message, 40))
+                                )
+
+        except socket.timeout as err:
+            failures.append('FAILURE: Port timed out\n' +
+                            '  Local port:    {}\n'.format(sock.getsockname()[1]) +
+                            '  Remote port:   {}\n'.format(process_port)
+                            )
+
+        except socket.error as err:
+            failures.append('FAILURE: Connection failed\n' +
+                            '  Local port:    {}\n'.format(sock.getsockname()[1]) +
+                            '  Remote port:   {}\n'.format(process_port) +
+                            '  Error message: {}\n'.format(err)
+                            )
+
+            # Reconnect
+            (failures, sock) = attempt_connect(failures, process_port, sock)
+            if sock is None:
+                break
+
+        tests_to_perform -= 1
+    
+    if sock is not None:
+        sock.shutdown(socket.SHUT_RDWR)
+        sock.close()
+
+    return failures
+
+# Remote device is setup to exit() if the first character of a message is 'a'
+def kill_remote_device():
+    try:
+        sock = socket.socket(socket.AF_INET, args.protocol)
+        sock.settimeout(10) # seconds
+        sock.connect((args.ip, start_port))
+        sock.send('a')
+        sock.close()
+
+    except socket.error as err:
+        # Do nothing and wait for remote device to timeout
+        print 'ERROR: Could not kill remote device'
+
+def test_iteration():
+    # Each process handles one remote port 
+    pool = Pool(args.remote_processes * args.remote_ports)
+
+    # Each process on the xC device has a pool of ports it can
+    # read from, with each pool spaced 10 apart
+    port_pool = [ j + (10*i) for i in range(0, args.remote_processes) for j in range(args.start_port, args.start_port + args.remote_ports)]
+    pool.map_async(process_test, port_pool, 1, callback=print_errors).get(9999999)
```

Comment:
It is very magic! It was used in development to deal with [this bug](https://bugs.python.org/issue8296), where it was impossible to kill the python subprocesses with a KeyboardInterrupt.